### PR TITLE
Pin versions of package in Dockerfile

### PR DESCRIPTION
**Pull request by fpesce**
As it is recommended by Dockerfile good practices (and reported by Dockerfile linter as DL3005 and DL3008).

We're going back and forth on this topic (should we pin, vs apt-get update / upgrade).
Full chain security might require that we pin in order to insure traceability of software versions shipped during a release. But we would need, during the release process to have a tool that would provide latest version of everything. It is a goal of a companion script `bin/dockerfile_versions_update.sh` that would make such suggestions.

The only alternative that I can see to that approach would be to update the version automatically during the execution of the Makefile for docker targets (using a slightly modified version of the script that would generate updated Dockerfile instead of advising updates), and then produce a manifest that should be shipped somewhere (the docker image itself?) for traceability purpose.

Let's continue that discussion here, @johnma14 & @sdake : any thoughts & suggestions?

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[X] Security
[X] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

**Comment by ericvn** [comment]
lgtm - I don't want to approve and have it go in, especially since the Title has WIP but I don't see a `do-not-merge` label.

**Comment by fpesce** [comment]
Removing the [WIP] , now it would be more a [Please Discuss] ;-)

**Comment by johnma14** [comment]
@fpesce what about the Dockerfiles for bookinfo sample apps? Shouldn't we be updating the package versions in there as well. I am fine with this approach of a script recommending upgrades as opposed to automatically updating the Dockerfiles for the reason that there could be dependencies where if you upgrade a version of a package it might break something else. For example, take a look at this PR:https://github.com/istio/istio/pull/16556/files. When we updated the `python` package version in `samples/bookinfo/src/productpage/Dockerfile`, we had to correspondingly update the `gevent` package version in `samples/bookinfo/src/productpage/requirements.txt` since the old gevent version was incompatible with the new python version we were upgrading to. Hope this helps. Thanks.

File: docker/Dockerfile.base [hunk 1]
```diff
@@ -8,22 +8,21 @@ ENV DEBIAN_FRONTEND=noninteractive
...
-      iputils-ping \
-      knot-dnsutils \
-      netcat \
-      tcpdump \
-      net-tools \
-      lsof \
-      linux-tools-generic \
-      sudo \
-   && update-ca-certificates \
-   && apt-get upgrade -y \
```
Comment (sdake) [inline docker/Dockerfile.base]: the problem with removing this line is that there is a whole slew of binaries in the base distro which (change) that are subject to CVEs.  We could just re-push the base image every release as that should pick up most CVEs, but I'm not sure that solves the manifest problem you want to solve.

Comment (johnma14) [reply docker/Dockerfile.base]: I agree @sdake.

**Comment by sdake** [comment]
@fpesce you may already be aware, but just in case you are not, Canonical (as well as most other distributions) push new base images every few weeks or more often.  Pinning the packages here works well enough, but doesn't necessarily pin transitive dependencies or distro packages that are used in the base image.  These may rev naturally without a CVE, or with an associated CVE.

File: docker/Dockerfile.base (hunk 1 above)
Comment (howardjohn) [reply docker/Dockerfile.base]: (I may be agreeing with you.. not sure) I don't think we want `upgrade`. See https://docs.docker.com/develop/develop-images/dockerfile_best-practices/. Instead, we should rely on the `ubunutu:bionic` image to be maintained and upgraded as appropriate? Although the 3 of you have spent much more time looking into this than me :slightly_smiling_face:

**Comment by sdake** [comment]
@fpesce I have ignored DL3008 across the project. If you really do wish to pin versions of distro packages, please file a bug and assign to me, and I'll revert https://github.com/istio/common-files/pull/254 and see what can be done about pinning.

The general problem with pinning of distro packages is that - on occasion - distro packages disappear from the APT / YUM / DNF repositories (those with CVEs are removed).

As we head into a world where distroless becomes feasible, the value of pinned packages makes sense (e.g. ca-certs).

@howardjohn - I don't recall what the exact solution was to the lack of debug tools images for K8s. Would you mind offering up the status of this feature in K8s - so we can transition the containers to distroless entirely and move away from packaged distros. We will probably continue to need packaged distros for some images - such as build-tools.

Cheers,
-steve

**Pull request closed without merging**